# Проект по SQL про книги

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.


Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

**Цель**

Анализ базы данных сервиса для чтения книг

**Задачи**

- Посчитайте, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитайте количество обзоров и среднюю оценку;
- Определите издательство, которое выпустило наибольшее число книг толще50 страниц — так вы исключите из анализа брошюры;
- Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
- Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.


**Содержании базы данных**

Таблица **books**


Содержит данные о книгах:
- book_id — идентификатор книги;
- author_id — идентификатор автора;
- title — название книги;
- num_pages — количество страниц;
- publication_date — дата публикации книги;
- publisher_id — идентификатор издателя.

Таблица **authors**


Содержит данные об авторах:
- author_id — идентификатор автора;
- author — имя автора.


Таблица **publishers**


Содержит данные об издательствах:
- publisher_id — идентификатор издательства;
- publisher — название издательства;


Таблица **ratings**


Содержит данные о пользовательских оценках книг:
- rating_id — идентификатор оценки;
- book_id — идентификатор книги;
- username — имя пользователя, оставившего оценку;
- rating — оценка книги.


Таблица **reviews**


Содержит данные о пользовательских обзорах на книги:
- review_id — идентификатор обзора;
- book_id — идентификатор книги;
- username — имя пользователя, написавшего обзор;
- text — текст обзора.

# Импортируем библиотеки и устанавливаем параметры подключения к базе данных

In [1]:
import pandas as pd
import sqlalchemy as sa

In [2]:
db_config = {
'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db' # название базы данных
}
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)
# сохраняем коннектор
engine = sa.create_engine(connection_string, connect_args={'sslmode':'require'})

Откроем все таблицы базы данных и посчитаем количество строк в таблицах

Таблица **books**

In [3]:
query = '''SELECT *
FROM books 
LIMIT 5;'''
con = engine.connect()
pd.read_sql(sql=sa.text(query), con = con)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [4]:
query = '''SELECT COUNT(*)
FROM books;'''
con = engine.connect()
pd.read_sql(sql=sa.text(query), con = con)

count
0   1000

Таблица **authors**

In [5]:
query = '''SELECT *
FROM authors
LIMIT 5;'''
con = engine.connect()
pd.read_sql(sql=sa.text(query), con = con)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [6]:
query = '''SELECT COUNT(*)
FROM authors;'''
con = engine.connect()
pd.read_sql(sql=sa.text(query), con = con)

count
0    636

Таблица **publishers**

In [7]:
query = '''SELECT *
FROM publishers
LIMIT 5;'''
con = engine.connect()
pd.read_sql(sql=sa.text(query), con = con)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [8]:
query = '''SELECT COUNT(*)
FROM publishers;'''
con = engine.connect()
pd.read_sql(sql=sa.text(query), con = con)

count
0    340

Таблица **ratings**

In [9]:
query = '''SELECT *
FROM ratings
LIMIT 5;'''
con = engine.connect()
pd.read_sql(sql=sa.text(query), con = con)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [10]:
query = '''SELECT COUNT(*)
FROM ratings;'''
con = engine.connect()
pd.read_sql(sql=sa.text(query), con = con)

count
0   6456

Таблица **reviews**

In [11]:
query = '''SELECT *
FROM reviews
LIMIT 5;'''
con = engine.connect()
pd.read_sql(sql=sa.text(query), con = con)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [12]:
query = '''SELECT COUNT(*)
FROM reviews;'''
con = engine.connect()
pd.read_sql(sql=sa.text(query), con = con)

count
0   2793

Вывела на экран первые 5 строк всех таблиц и посчитала количество строк в кадой:
1. Таблица books - 1000 строк
2. Таблица authors - 636 строк
3. Таблица publishers - 340 строк 
4. Таблица ratings - 6456 строк
5. Таблица reviews - 2793 строк

## Посчитайте, сколько книг вышло после 1 января 2000 года

In [13]:
query = '''SELECT COUNT(book_id)
FROM books 
WHERE publication_date > '2000-01-01';'''
con = engine.connect()
pd.read_sql(sql=sa.text(query), con = con)

count
0    819

819 книг вышло после 1 января 2000 года

## Для каждой книги посчитайте количество обзоров и среднюю оценку

In [14]:
query ='''
SELECT b.book_id,
       b.title,
       COUNT(DISTINCT rev.review_id) AS reviews_count,
       AVG(rat.rating) AS avg_rating
FROM books as b
  LEFT JOIN reviews AS rev ON b.book_id = rev.book_id
  LEFT JOIN ratings AS rat ON b.book_id = rat.book_id
GROUP BY
  b.book_id
ORDER BY
  reviews_count DESC, avg_rating DESC;'''

    
con = engine.connect()
pd.read_sql(sql=sa.text(query), con = con)

book_id                                              title  \
0        948                            Twilight (Twilight  #1)   
1        302  Harry Potter and the Prisoner of Azkaban (Harr...   
2        299  Harry Potter and the Chamber of Secrets (Harry...   
3        656                                     The Book Thief   
4        734                                   The Glass Castle   
..       ...                                                ...   
995      191  Disney's Beauty and the Beast (A Little Golden...   
996      387                               Leonardo's Notebooks   
997      221                          Essential Tales and Poems   
998       83    Anne Rice's The Vampire Lestat: A Graphic Novel   
999      808                            The Natural Way to Draw   

     reviews_count  avg_rating  
0                7    3.662500  
1                6    4.414634  
2                6    4.287500  
3                6    4.264151  
4                6    4.206897  
..             ...         ...  
995              0    4.000000  
996              0    4.000000  
997              0    4.000000  
998              0    3.666667  
999              0    3.000000  

[1000 rows x 4 columns]

Есть книги, у которых нет оценок, но есть рейтинг. Максимальная оценка 7 у книги Twilight (Twilight #1).

## Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры

In [15]:
query ='''
SELECT publisher,
    COUNT(b.book_id) AS count_books
FROM publishers as p
  JOIN books AS b ON p.publisher_id = b.publisher_id
WHERE b.num_pages > 50
GROUP BY p.publisher
ORDER BY count_books DESC
LIMIT 1;'''

    
con = engine.connect()
pd.read_sql(sql=sa.text(query), con = con)

publisher  count_books
0  Penguin Books           42

Издательство, которое выпустило больше всего книг толще 50 страниц - Penguin Books

## Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

In [16]:
query ='''
WITH temp AS (SELECT a.author,
              AVG(r.rating) AS avg_rating
FROM author as a
  JOIN books AS b ON b.author_id = a.author_id
  JOIN ratings AS r ON b.book_id = r.book_id
GROUP BY b.book_id, a.author
HAVING COUNT(r.rating_id)>=50 
ORDER BY avg_rating DESC)
SELECT author,
       ROUND(AVG(avg_rating),2) AS rating_avg
FROM temp
GROUP BY author
ORDER BY rating_avg DESC
LIMIT 1;'''

    
con = engine.connect()
pd.read_sql(sql=sa.text(query), con = con)

author  rating_avg
0  J.K. Rowling/Mary GrandPré        4.28

Автора с самой высокой средней оценкой книг — J.K. Rowling/Mary GrandPré

## Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок

In [17]:
query ='''
WITH rating AS (SELECT rat.username,
                        COUNT(rating_id) AS rat_count
                FROM ratings AS rat
                GROUP BY rat.username
                HAVING COUNT(rating_id) > 48),
    review AS (SELECT rev.username,
                      COUNT(review_id) AS rev_count
               FROM reviews rev
               GROUP BY rev.username)
SELECT
    AVG(rev_count) as avg_rev
FROM rating
    JOIN review ON review.username = rating.username '''

    
con = engine.connect()
pd.read_sql(sql=sa.text(query), con = con)

avg_rev
0     24.0

Cреднее количество обзоров от пользователей - 24

**Вывод:**

В приложении загруженно больше 819 книг, которые имеют оценки и рейтинги. Есть книги, у которых нет оценок, но есть рейтинг, можно подобрать для себя чтиво основываяст на любой из показателей. Максимальная оценка 7 у книги Twilight (Twilight #1). Есть книги крупных издательств, которые выпускают большое количество книг всего книг толщиной в 50 страниц и больше. Например, лидирующее издательство - Penguin Books, количество их толстых книг в приложении - 42. Также есть и популярные писатели с лучшими рейтингами, например, J.K. Rowling/Mary GrandPré. Еще в приложении можно написать или прочитать обзоры, они помогут выбрать интереную книгу. Cреднее количество обзоров от пользователей - 24